db/pebble/transaction.go

```diff
@@ -18,25 +18,30 @@ type Transaction struct {
 }
 
 // Discard : see db.Transaction.Discard
-func (t *Transaction) Discard() (err error) {
+func (t *Transaction) Discard() error {
 	if t.batch != nil {
-		err = t.batch.Close()
+		if err := t.batch.Close(); err != nil {
+			return err
+		}
 		t.batch = nil
```

Comment:
This changes discard behavior on failure. batch & snapshot wont be set to nil if close fails. It might be better this way, let's see what others think.

Reply:
The behaviour should remain the same because we are using nil to enforce calling Discard multiple times on the same transaction does nothing. If we don't set batch and snapshot to nil we guarantee second and subsequent calls to discard would error. Also, we risk causing deadlocks as we may not call unlock on line 36 when we need to.

Reply:
It lets you to retry when Discard returns an error tho., that is why I thought it might be better this way.

Reply:
I think for retying we would require more information and it will have to depend on the error that pebble returns. This would be difficult to determine which errors are recoverable and which errors are not recoverable. Therefore, I would like the behaviour to remain the same.

If the behaviour changes then we need to update the tests accordingly.

Reply:
> It would be difficult to determine which errors are recoverable and which errors are not. 

that is not a concern of the pebble package tho. That responsibility lies on the caller.

Question here is do we give user the ability to retry or not. My 2 cents is that I think we should.

Reply:
> having a dangling snapshot/batch open on the database which we have no reference to is not a pleasant situation to be in.

This is a good point. 

My concern was how would the caller decide that they need to retry without switching on a specific error.